R/infix_spaces_linter.R

```diff
@@ -1,43 +1,81 @@
+  infix_metadata$xml_tag
 )
-binary_infix_tokens <- c(
-
-  "GT" = "GT",                     # >        : greater than
-  "GE" = "GE",                     # <=       : greater than or equal to
-  "LT" = "LT",                     # <        : less than
-  "LE" = "LE",                     # <=       : less than or equal to
-  "EQ" = "EQ",                     # ==       : vector equality
-  "NE" = "NE",                     # !=       : not equal
-  "AND" = "AND",                   # &        : vector boolean and
-  "OR" = "OR",                     # |        : vector boolean or
-  "AND2" = "AND2",                 # &&       : scalar boolean and
-  "OR2" = "OR2",                   # ||       : scalar boolean or
-  "LEFT_ASSIGN" = "LEFT_ASSIGN",   # <- or := : left assignment
-  "RIGHT_ASSIGN" = "RIGHT_ASSIGN", # ->       : right assignment
-  "EQ_ASSIGN" = "EQ_ASSIGN",       # =        : equal assignment
-  "EQ_SUB" = "EQ_SUB",             # =        : keyword assignment
-  "SPECIAL" = "SPECIAL",           # %[^%]*%  : infix operators
-  "OP-SLASH" = "'/'",              # /        : unary division
-  "OP-STAR" = "'*'",               # *        : unary multiplication
-
-  NULL
+# treated separately because spacing rules are different for unary operators
+infix_metadata$unary <- infix_metadata$xml_tag %in% c("OP-PLUS", "OP-MINUS", "OP-TILDE")
+# high-precedence operators are ignored by this linter; see
+#   https://style.tidyverse.org/syntax.html#infix-operators
+infix_metadata$low_precedence <- infix_metadata$string_value %in% c(
+  "+", "-", "~", ">", ">=", "<", "<=", "==", "!=", "&", "&&", "|", "||", "<-", "->", "=", "%%", "/", "*"
 )
-infix_tokens <- c(unary_infix_tokens, binary_infix_tokens)
 
 #' Infix spaces linter
 #'
-#' Check that infix operators are surrounded by spaces.
+#' Check that infix operators are surrounded by spaces. Enforces the corresponding Tidyverse style guide rule;
+#'   see <https://style.tidyverse.org/syntax.html#infix-operators>.
 #'
+#' @param exclude_operators Character vector of operators to exlude from consideration for linting.
+#'   Default is to include the following "low-precedence" operators:
+#'   `+`, `-`, `~`, `>`, `>=`, `<`, `<=`, `==`, `!=`, `&`, `&&`, `|`, `||`, `<-`, `:=`, `<<-`, `->`, `->>`,
+#'   `=`, `/`, `*`, and any infix operator (exclude infixes by passing `"%%"`). Note that `<-`, `:=`, and `<<-`
+#'   are included/excluded as a group (indicated by passing `"<-"`), as are `->` and `->>` (_viz_, `"->"`),
+#'   and that `=` for assignment and for setting arguments in calls are treated the same.
 #' @evalRd rd_tags("infix_spaces_linter")
 #' @seealso [linters] for a complete list of linters available in lintr.
 #' @export
-infix_spaces_linter <- function() {
+infix_spaces_linter <- function(exclude_operators = NULL) {
   Linter(function(source_file) {
+    infix_tokens <- infix_metadata[
+      with(infix_metadata, low_precedence & !string_value %in% exclude_operators),
```

Comment:
no visible binding for global variable ‘string_value’ lint (could be fixed via global variables or by defining them as `NULL` in the function scope)

Reply:
I think this is a false positive in the linter since `R CMD check` doesn't complain about it.

Reply:
Filed #941 as follow-up